### DISCUSSION TITLE
7 novembre 2018 - esercizio 3

### DISCUSSION THREAD

**Discussion by AglaiaNorza** [sapienzastudentsnetwork/basi-di-dati-1#98] [Esami]
E' dato un file di 134.800 record. Ogni record occupa 257 byte, di cui 57 per la chiave. Un blocco contiene 2048 byte.

Un puntatore a blocco occupa 4 byte. Si utilizza una organizzazione hash con record distribuiti uniformemente tra 250 bucket.

3a) Calcolare l'occupazione della bucket directory e dei bucket

3b) Calcolare il costo medio di una ricerca considerando chiavi di hash univoche

3c) Quanti bucket occorrerebbero per avere un costo medio di ricerca minore o uguale a 20 accessi?

**Comment by sprigatita** [comment]
Mia personale proposta
![image](https://github.com/user-attachments/assets/84c717da-46d9-4b5b-b9bf-2ca8e22cf8df)
![image](https://github.com/user-attachments/assets/40d9d683-2474-4d7e-a807-4cb6f183d147)

**Comment by FeddyLix17** [comment]
3a)

La bucket directory dovrà contenere un puntatore per ogni bucket

$$\Large Spazio\ totale\ necessario = numero\ bucket * dimensione\ puntatore =$$

$$\Large = 250 * 4\ byte = 1000\ byte = 1KB$$


$$\Large Blocchi\ necessari\ per\ bucket\ directory = \left\lceil\frac{1000}{2048}\right\rceil = \lceil0.488\rceil = 1\ blocco$$

Data la distribuzione uniforme di questi ultimi, in ogni bucket ci saranno circa

$$\Large Record\ per\ bucket = \left\lceil\frac{134.800}{250}\right\rceil = 540\ record$$

In un singolo blocco, entreranno circa

$$\Large Record\ per\ blocco = \left\lfloor\frac{2048}{257}\right\rfloor = \left\lfloor 7.968\right\rfloor = 7\ record$$

$$\Large Blocchi\ necessari\ per\ singolo\ bucket\= \left\lceil\frac{540}{7}\right\rceil = \lceil77.14\rceil = 78\ blocchi$$

---

3b)

Oltre al singolo accesso alla bucket directory, data sempre la loro distribuzione uniforme, ogni record sarà [equiprobabilmente](https://math.stackexchange.com/questions/4152808/why-is-the-expected-value-for-ourcomes-of-equal-probabilities-different-per-outc) presente in uno (e uno solo) dei blocchi presenti in un singolo bucket.

Dal punto precedente sappiamo che un bucket conterrà 78 blocchi, da cui

$$\Large Accessi\ medi\ in\ un\ singolo\ bucket = \left\lfloor\frac{78}{2}\right\rfloor = 39\ accessi$$

$$\Large Costo\ medio\ ricerca\ di\ un\ record = 39 + 1 = 40\ accessi$$

---

3c)

Impostando l'equazione 

$$\Large Accesso\ Bucket\ Directory + Accessi\ medi\ in\ un\ singolo\ bucket \leq 20$$

$$\Large 1 + Accessi\ medi\ in\ un\ singolo\ bucket \leq 20$$

$$\Large Accessi\ medi\ in\ un\ singolo\ bucket < 19$$

Ricordando che gli Accessi medi in un singolo bucket, in questo caso, sono uguali a $\Large \frac{blocchi\ per\ bucket}{2}$

$$\Large \frac{blocchi\ per\ bucket}{2} \leq 19 \implies blocchi\ per\ bucket < 38$$

Se in questo caso un bucket dovrà contenere al più 38 blocchi 

dunque un bucket conterrà al più 38 * 7 = 266 record

$$\Large Record\ massimi\ per\ bucket = numero\ massimo\ blocchi * record\ per\ blocco =$$

$$\Large = 38 * 7 = 266\ record$$

$$\Large Numero\ minimo\ bucket\ necessari = \left\lceil\frac{134.800}{266}\right\rceil = \left\lceil506.76\right\rceil = 507\ bucket$$
